server/player_service.py

```diff
@@ -167,20 +179,33 @@ async def _fetch_player_legacy_rating(self, player, rating_types, conn):
                 )
                 continue
 
-
             player.ratings[rating_type] = (
                 row[table.format("mean")], row[table.format("deviation")]
             )
             player.game_count[rating_type] = row[table.format("numGames")]
 
-
     def remove_player(self, player: Player):
         if player.id in self._players:
             del self._players[player.id]
             metrics.players_online.set(len(self._players))
 
-    def get_permission_group(self, user_id: int) -> int:
-        return self.privileged_users.get(user_id, 0)
+    async def has_permission_role(self, player: Player, role_name: str) -> bool:
+        async with self._db.acquire() as conn:
+            result = await conn.execute(
+                select([group_permission_assignment.c.group_id])
+                .select_from(
+                    group_permission
+                    .join(group_permission_assignment, onclause=(
+                        group_permission.c.id ==
+                        group_permission_assignment.c.permission_id
+                    ))
+                )
+                .where(group_permission.c.technical_name == role_name)
+            )
+            async for row in result:
+                if row.group_id in player.user_groups:
+                    return True
+            return False
```

Comment:
As I mentioned before, you don't need to have `player.user_groups`, because you're making another query to the database here. just join proper tables and get rid of `player.user_groups`. It's not hard for MySQL to have one more join.

What you will check is only that you have a row, so you can select `text(1).label("right_exists"). .... .first()`.

In the end, fewer queries to the database, and less code to maintain.

Reply:
It's still the same number of queries because we need the technical_names for `Player.is_moderator()`. It's just that the way I have it, we use 1 fewer join.

Reply:
yeah, np.